### DISCUSSION TITLE
Confused on what exactly are the differences that come from wrapping react-query

### DISCUSSION THREAD

**Discussion by cysabi** [trpc/trpc#2932] [Q&A]
Hello, I've been trying to implement this library and, while reading the documentation I've gotten myself quite confused over exactly how react-query is wrapped.

# query keys
![image](https://user-images.githubusercontent.com/47790183/194452106-cf3b7fcd-54d6-4581-a1eb-08e77886fc02.png)

This seems to state that the only options that are exposed from `useQuery` is the input field, and then a react-query query options field (plus an ssr and abortOnUnmount field)

![image](https://user-images.githubusercontent.com/47790183/194452354-4df13b85-b80f-4656-80ec-078414876cd5.png)

This seems to imply a `byId` query that has a query key of `['post', 'byId', {id: input.id}]`

However, I couldn't find any explicit documentation on how query keys are handled. At first I thought it was something you can explicitly append when calling `useQuery` hook, but as shown above, there's only 2 fields, the input field and the options field. The react-query documentation also does not have any way to append id in it's options.

![image](https://user-images.githubusercontent.com/47790183/194452673-d0f5cecf-8688-4993-8666-a86625739574.png)

This seems to imply that somehow `{id: 1}` is being automatically appended to the query key when you call `useQuery` which isn't documented anywhere. 

And it wouldn't make much sense even if it was true, as `{id: 1}` is the **input parameter**. What if someone puts a whole user profile object in that input parameter, there's no reason that should be appended to the query key instead of just the id.

At the same time, there should absolutely be a way for users to append ids to the query key array, as otherwise it can result in a lot of unnecessary invalidating and refetching from react-query.

Am I missing something? Am I blind and just haven't seen the explicit documentation on handling query keys and etc.?

# query client
So I have experience with react-query, and I expected `createTRPCNext` to expose the query client variable to invalidate queries etc. But instead, in order to invalidate queries, you have to use `trpc.myroute.invalidate()`

While this definitely caught me off guard, it completely makes sense, as query key is abstracted away from you, and you already have objects for each query.

However, it's the only (documented) function that wraps one of the query client's features:
![image](https://user-images.githubusercontent.com/47790183/194453202-cb8fb8ac-2f78-41f6-b39b-3d7a90375ba0.png)

There's no documented `trpc.myroute.refetch()` or `trpc.myroute.prefetch()`. So you have to use the `useQueryClient` hook *anyway* if you want to use anything except `invalidateQueries`. But you're never explicitly told how the query key is calculated, so how do you know what to use? (without cross referencing the devtools and your own experimentation)

**Comment by cysabi** [comment]
> **UPDATE:**

After inferring based on the page that documents static-site-generation helpers, I think there are prefetch functions:
![image](https://user-images.githubusercontent.com/47790183/194458498-6520611e-a025-4534-8c66-58ee4c514bb8.png)

However, I still haven't found a full documented list of all the converted functions, for all I know it could only be those 2.

**Reply by KATT** [reply]
Hey, it's true the query key isn't documented & that we don't have *all* functions that react query has.

**Comment by iway1** [comment]
> What if someone puts a whole user profile object in that input parameter, there's no reason that should be appended to the query key instead of just the id.

anything you pass to a query as input *should* affect what data gets returned (otherwise why is it being passed to the end point?). therefor it makes sense to always append the input to the query key. This implementation actually enforces good react-query usage, there's no reason to pass things that aren't used to calculate the return data and there's no reason to use things that aren't passed as query keys.

**Reply by cysabi** [reply]
Thanks for the explanation, this makes a lot of sense!
the only thing you should pass in the input parameter is some sort of extra key.
I sort of realized this a while after I posted it, but I still find it a bit confusing if only due to the fact that there's no documentation on how query keys work (and how they should be used) at all.
